CI: Sign and package Windows builds

derrod opened the pull request:
### Description

This adds a new workflow that will create signed Windows builds and installer on tag pushes.

To facilitate code signing in a way that complies with the new requirements to use HSM (Hardware Security Modules) for key storage[^1] we utilise [Google Cloud's Key Management Service](https://cloud.google.com/security-key-management).

For signing Windows binaries we continue to use `signtool` with Google's CNG Provider that allows it to run the signing operation on the cloud HSM: https://cloud.google.com/kms/docs/reference/cng-signtool

#### ToDo:
- [ ] Actually acquire new code signing certificate **(In progress)**
  + Must be ECDSA, Google's CNG provider does not support RSA keys (yet): https://github.com/GoogleCloudPlatform/kms-integrations/issues/16
  + GlobalSign will work according to https://icedev.pl/posts/setting-up-ev-code-signing-google-hsm-fips-140-2/
  + DigiCert is probably also fine based on responses in issue linked above
  + If we wish to import the current keys to migrate before acquiring a new cert we will probably have to use [jsign](https://github.com/ebourg/jsign)
- [x] Figure out if we want to do delta patch generation elsewhere (takes a long time + storage on GitHub CI)
  + **Resolution:** It's not actually to bad (<20 minutes, ~600 MiB per version) as long as there's only one CEF version change to cover
- [x] Maybe find a better way for downloading old builds + PDBs for delta generation/packaging (GCS? S3? SFTP?)
  + **Resolution:** Google Cloud Storage (GCS) seems like a good option and we're pretty much going to stay within the free tier
- [x] #9816 

[^1]: https://knowledge.digicert.com/general-information/new-private-key-storage-requirement-for-standard-code-signing-certificates-november-2022

### Motivation and Context

Reduces bus factor a bit.

### How Has This Been Tested?

Tested on my fork with test cert and keys: https://github.com/derrod/obs-studio/actions/runs/6817205390/job/18540744248

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,122 @@
+        # Extract artifact in build dir and remove zip
+        cd build
+        Expand-Archive -Path *.zip -DestinationPath .
+        del *.zip
```
I'd suggest to use `Remove-Item` - our PowerShell code mostly tries to keep using first-class PowerShell idioms as much as possible, only falling back on program invocations if necessary.

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,122 @@
+        BOUF_NSIS_HASH: 'bb0561a76cecaba62785642dd71052a9fe53458e45df108ca08867ee333b80bd'
+      run: |
+        # Download bouf release
+        mkdir bouf && cd bouf
```
Copy the `Ensure-Location` PowerShell module from the GitHub build scripts to the GitHub action and source it from there.

derrod replied:
The repo is checked out in this case so I can just import it rather than having to copy it, no? I don't think they'd be imported automatically.

PatTheMav replied:
Yeah both work, I’d just prefer having them contained as that would make extracting them into a separate repo easier if we chose to do that.

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,122 @@
+        # Download bouf release
+        mkdir bouf && cd bouf
+        
+        curl -o bouf-windows.zip -L https://github.com/obsproject/bouf/releases/download/$Env:BOUF_VERSION/bouf-windows-$Env:BOUF_VERSION.zip
```
Copy the `Invoke-External` PowerShell module from the GitHub build scripts to the GitHub action and source it from there.

Also enables proper error handling whenever a CLI application fails to run.

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,122 @@
+        RCLONE_DRIVE_ENV_AUTH: 'true'
+        RCLONE_DRIVE_SHARED_WITH_ME: 'true'
+      run: |
+        rclone copy --transfers 40 :drive:old_builds "${{ github.workspace }}/old_builds"
```
Shouldn't `robocopy` fulfil that need? Especially for local file operations, `rclone` as a third-party library seems unnecessary.

derrod replied:
rclone is used to download data from a remote storage system (google cloud in this case), robocopy is just for local copies.

PatTheMav replied:
Ah my mistake, that's fine then.

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,122 @@
+      env:
+        UPDATER_PRIVATE_KEY: ${{ inputs.updaterPrivateKey }}
+      run: |
+        &"${{ github.workspace }}\bouf\bin\bouf.exe" `
```
Use splatting - put all arguments into an array, then just invoke via `@<array>`. That also avoids the use of the backtick escaping of newlines.

derrod replied:
I don't really see an advantage to that, in fact, it would require more backticks to escape all the `"`!

RytoEX replied:
I don't think there's any extra escaping? PowerShell should do that for you. Something like this @PatTheMav ?
```pwsh
$boufOptions = @(
    '--config', "${{ github.workspace }}/repo/build-aux/bouf/config.toml"
    '--version', "${{ inputs.version }}"
    '--branch', "${{ inputs.channel }}"
    '--notes-file', "${{ github.workspace }}/notes.rst"
    '-i', "${{ github.workspace }}/build"
    '-p', "${{ github.workspace }}/old_builds"
    '-o', "${{ github.workspace }}/output"
)

Invoke-External "${{ github.workspace }}\bouf\bin\bouf.exe" @boufOptions
```

derrod replied:
I suppose that works, I was thinking of having to put flag and value into one but bouf doesn't use the `--flag=<value>` form.

PatTheMav commented on .github/workflows/push.yaml:
```diff
@@ -194,6 +194,58 @@ jobs:
+        shell: bash
+        run: |
+          channel='stable'
+          if [[ ${GITHUB_REF_NAME} =~ (beta|rc) ]]; then
```
Aren't we always escaping the first token to the test operation in Bash? IIRC it's not required in ZSH but in Bash you should still always double-quote the first token to avoid automatic tokenisation (if the variable contains spaces the effective code would be `[[ some value =~ (beta|rc) ]]` and that's invalid shell code.

derrod replied:
I think that is correct, though git refs shouldn't be able to have spaces. I haven't bothered rewriting this to powershell yet, so it was mostly just a stripped down version of what's used for the appcast.

PatTheMav commented on build-aux/bouf/cng-release-signing-key.pem:
```diff
@@ -0,0 +1,5 @@
+MHYwEAYHKoZIzj0CAQYFK4EEACIDYgAEtfLbXkHUVc9oUPTNyaEK3hIwmuGRoTtd
+6zDhwqjJuYaMwNd1aaFQLMawTwZgR0Xn27ymVWtqJHBe0FU9BPIQ+SFmKw+9jSwu
+/FuqbJnLmTnWMJ1jRCtyHNZawvv2wbiB
+-----END PUBLIC KEY-----
```
Missing newline at the end of the file.

PatTheMav commented on build-aux/bouf/config.toml:
Is this a generic `bouf` config file that should be used locally as well? If it's only used for the GitHub action, it should be moved into the action's directory.

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+    description: Google service account name
+    required: true
+  version:
+    description: Version string (e.g. 30.0.0-rc1)
```
```suggestion
    description: Version string (e.g., 30.0.0-rc1)
```

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+    - name: Extract Artifact
+      shell: pwsh
+      run: |
+        Expand-Archive -Path build\*.zip -DestinationPath build
+        Remove-Item build\*.zip
```
Did the asterisks actually need to be escaped?

PatTheMav replied:
I think it's matching files in the `build` subdirectory.

RytoEX replied:
> I think it's matching files in the `build` subdirectory.

Oh, that makes more sense.

derrod replied:
Yep, remember Windows supports both `/` and `\` but `\` is technically correct (and sometimes built-in stuff won't work correctly with `/`)

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+        Expand-Archive -Path *.zip
+        $sigPath = Get-ChildItem *.sig -Recurse
+        $msiPath = Get-ChildItem *.msi -Recurse
+        # Verify digital signature against Google's public key
+        Invoke-External openssl dgst -sha384 -verify "$env:GITHUB_ACTION_PATH/cng-release-signing-key.pem" -signature $sigPath $msiPath
```
I suppose if we're verifying the digital signature, we don't need to do a hash check?

derrod replied:
Indeed.

RytoEX replied:
Fair enough.

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+        # Finally, install the CNG provider
+        Invoke-External msiexec /i $msiPath /qn /norestart
+
+    - name: Install rclone and pandoc
```
```suggestion
    - name: Install pandoc and rclone
```

Super nit. Just bothers me.

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+
+    - name: Install rclone and pandoc
+      shell: pwsh
+      run: choco install pandoc rclone --no-progress -y
```
I'd have suggested winget for this, except pandoc is a bit outdated (3.1.1 on winget, 3.1.9 on Chocolatey) and rclone is not yet available on winget. Something to keep in mind for the future if it reaches parity maybe.

derrod replied:
Winget isn't available on runner as far as I could tell

RytoEX replied:
Ah, fair point.

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+      shell: pwsh
+      run: |
+        # Release notes are just the tag body on Windows
+        cd repo
```
```suggestion
        Ensure-Location repo
```
Perhaps?

derrod replied:
If repo doesn't exist this should fail rather than creating it

PatTheMav replied:
That'd be `Set-Location` then.

derrod replied:
`cd` is an alias for `Set-Location anyway`:
```pwsh
PS C:\Dev\obs\obs-studio> Set-Location asd
Set-Location: Cannot find path 'C:\Dev\obs\obs-studio\asd' because it does not exist.
PS C:\Dev\obs\obs-studio> cd asd
Set-Location: Cannot find path 'C:\Dev\obs\obs-studio\asd' because it does not exist.
```

PatTheMav replied:
Indeed, but at least for use in scripts I prefer us to use the "long form" of PowerShell commands, at the very least for consistency with our other PowerShell code.

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+        workload_identity_provider: ${{ inputs.gcpWorkloadIdentityProvider }}
+        service_account: ${{ inputs.serviceAccountName }}
+
+    - name: Download old builds
```
```suggestion
    - name: Download Old Builds
```

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+           "-p",           "${{ github.workspace }}/old_builds"
+           "-o",           "${{ github.workspace }}/output"
+        )
+        &"${{ github.workspace }}\bouf\bin\bouf.exe" @boufArgs
```
```suggestion
        Invoke-External "${{ github.workspace }}\bouf\bin\bouf.exe" @boufArgs
```

If not, then at least:
```
& "${{ github.workspace }}\bouf\bin\bouf.exe" @boufArgs
```

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+        )
+        &"${{ github.workspace }}\bouf\bin\bouf.exe" @boufArgs
+
+    - name: Cleanup bouf output dir
```
```suggestion
    - name: Cleanup bouf Output Directory
```

RytoEX commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,127 @@
+    - name: Cleanup bouf output dir
+      shell: pwsh
+      run: |
+        # Compress updater data so github's artifact upload won't take forever
```
```suggestion
        # Compress updater data so GitHub's artifact upload won't take forever
```

PatTheMav commented on .github/actions/bouf/Ensure-Location.ps1:
```diff
@@ -0,0 +1,29 @@
+function Ensure-Location {
```
Correct - in theory all the utility functions could be put into a single imported PowerShell script used just by the action (instead of the separate files used by the build script), but that's more a matter of style/taste.

RytoEX replied:
> Correct - in theory all the utility functions could be put into a single imported PowerShell script used just by the action (instead of the separate files used by the build script), but that's more a matter of style/taste.

Maybe some day it might make sense to turn the utility scripts into modules and install them as needed. Not a today problem.

If this particular method is okay to you, then I'll not block on it. We'll just need to be cognizant that there are multiple copies floating around if we have any changes to make.

derrod replied:
I suppose the files here could be symlinks rather than direct copies, but not sure how GitHub would handle that for actions.

RytoEX replied:
It's probably fine as-is for now. Just need to remember to port any fixes.

derrod commented:
Updated the actions a bit:
- All the files in the `actions/bouf/` are now hash-checked before running the action
- Pinned specific versions for pandoc and rclone
- Pinned specific commits for third-party actions (gcloud auth + sdk setup)
- Switched to using GCS (Google Cloud Storage) for old builds
- Signing the updater manifest is now also done via Cloud KMS using the `gcloud` CLI tool
  + This didn't seem to work via `Invoke-External` as it passes the arguments in an incompatible way, haven't investigated further but shouldn't matter, this failing isn't critical either

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,137 @@
+        Ensure-Location bouf
+        Invoke-External gh release download "$env:BOUF_TAG" -R "obsproject/bouf" -p "*windows*.zip" -p "*nsis*.zip"
+        
+        if ((Get-FileHash *windows*.zip -Algorithm SHA256).Hash -ne "$env:BOUF_HASH") {
```
This will work, though unexpectedly - the `gh` glob pattern would potentially download multiple files, and `Get-FileHash <pattern>).Hash` would yield multiple hashes (one per line) then. The `-ne` expression in turn yields all the hashes that do not match the comparator (so not a `$true` or `$false` boolean), but the string will be interpreted as `$true` so the intended effect is the same.

Too broad star expressions in a glob are icky, but should be fine for now.

derrod replied:
Considering that this is one of our repos (well, mostly mine) I don't expect there to ever be more than one artifact per pattern as there's only one windows build or NSIS zip.

PatTheMav replied:
Correct, which is why I'm not concerned. I personally just prefer code to be "hardened" logically to not rely on assumptions. Doesn't require to be changed though.

derrod replied:
I just updated it now to use the same pattern as the creation of those zips.

PatTheMav commented on .github/workflows/push.yaml:
```diff
@@ -194,6 +194,71 @@ jobs:
+          BOUF_ACTION_HASH: '63933179a2cc636cb2bca2ae4ac2ae30af71bf29cc2a71f6ab01600a6f586bfe'
+        run: |
+          $shortHash = $env:GITHUB_SHA.Substring(0,9)
+          $channel = 'stable'
```
Use PowerShell conditional expression:

```
$output = if ($env:GITHUB_REF_NAME -match "(beta|rc)") { "beta" } else { "stable" }
```

It's similar to pattern matching expressions.

RytoEX commented on .github/workflows/push.yaml:
```diff
@@ -194,6 +194,68 @@ jobs:
+          
+          # Ensure files in action haven't been modified
+          $folderHash = ''
+          $files = Get-ChildItem "${{ github.workspace }}\repo\.github\actions\bouf"
+          foreach ($file in $files) {
+            $folderHash += (Get-FileHash $file -Algorithm SHA256).Hash
+          }
+          # This is stupid but so is powershell
+          $stream = [IO.MemoryStream]::new([byte[]][char[]]$folderHash)
+          if ((Get-FileHash -InputStream $stream -Algorithm SHA256).Hash -ne "$env:BOUF_ACTION_HASH") {
+            throw "bouf action folder hash does not match."
+          }
```
Could you walk me through the intent here?

derrod replied:
Ensure that all files in the `.github/actions/bouf` folder match an expected hash to verify they have not been modified. For that purpose they are hashed, those hashes are concatenated, and the hash of the concatenated hash is verified against a known one. That's a little simpler than maintaining a file or environment variables with hashes for each file individually.

RytoEX replied:
Ah, so it's the "hash of the concatenated hash" that is why you needed a MemoryStream. Okay.

derrod replied:
Unfortunately, despite https://github.com/PowerShell/PowerShell-RFC/issues/65 being merged over 6 years ago PowerShell still has no native means of hashing a string.

PatTheMav commented on .github/workflows/push.yaml:
```diff
@@ -194,6 +194,68 @@ jobs:
           path: ${{ github.workspace }}/output
 
+  create-windows-update:
+    name: Create Windows Update 🥩
```
Heh.

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,139 @@
+        . $env:GITHUB_ACTION_PATH\Ensure-Location.ps1
+        . $env:GITHUB_ACTION_PATH\Invoke-External.ps1
+        Ensure-Location bouf
+        $windows_zip = "bouf-windows-$env:BOUF_TAG.zip"
```
We usually enclose variables for interpolation with braces (similar to Bash) to ensure the variable name is clearly scoped (e.g. `${env:BOUF_TAG}` or `$($Object.Method())`).

Outside of interpolation there is usually no need for enclosing (or even quoting in some cases).

PatTheMav commented:
Overall seems fine to me apart from PowerShell code style nits.

RytoEX reviewed (approved):
Seems fine.

derrod commented:
Updated to use `actions/upload-artifact@v4` instead of `v3` which massively increases the speed of uploads. Also the outputs are now separated in a logical manner (installer, signed ZIP, updater files, PDBs) rather than all in one.

RytoEX reviewed (approved):
Seems fine again to me.

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,133 @@
+        Ensure-Location bouf
+        $windows_zip = "bouf-windows-${env:BOUF_TAG}.zip"
+        $nsis_zip = "bouf-nsis-${env:BOUF_TAG}.zip"
+        Invoke-External gh release download "$env:BOUF_TAG" -R "obsproject/bouf" -p $windows_zip -p $nsis_zip
```
```suggestion
        Invoke-External gh release download "${env:BOUF_TAG}" -R "obsproject/bouf" -p $windows_zip -p $nsis_zip
```

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,133 @@
+        $nsis_zip = "bouf-nsis-${env:BOUF_TAG}.zip"
+        Invoke-External gh release download "$env:BOUF_TAG" -R "obsproject/bouf" -p $windows_zip -p $nsis_zip
+        
+        if ((Get-FileHash $windows_zip -Algorithm SHA256).Hash -ne "$env:BOUF_HASH") {
```
```suggestion
        if ((Get-FileHash $windows_zip -Algorithm SHA256).Hash -ne "${env:BOUF_HASH}") {
```

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,133 @@
+        if ((Get-FileHash $windows_zip -Algorithm SHA256).Hash -ne "$env:BOUF_HASH") {
+          throw "bouf hash does not match."
+        }
+        if ((Get-FileHash $nsis_zip -Algorithm SHA256).Hash -ne "$env:BOUF_NSIS_HASH") {
```
```suggestion
        if ((Get-FileHash $nsis_zip -Algorithm SHA256).Hash -ne "${env:BOUF_NSIS_HASH}") {
```

PatTheMav commented on .github/actions/bouf/action.yaml:
```diff
@@ -0,0 +1,133 @@
+        . ${env:GITHUB_ACTION_PATH}\Invoke-External.ps1
+        Ensure-Location gcng
+        
+        Invoke-External gh release download "$env:CNG_TAG" -R "GoogleCloudPlatform/kms-integrations" -p "*amd64.zip"
```
```suggestion
        Invoke-External gh release download "${env:CNG_TAG}" -R "GoogleCloudPlatform/kms-integrations" -p "*amd64.zip"
```
